### PR TITLE
#169817553 Add approving or rejecting request

### PR DESCRIPTION
**Pull request by gadishimwe**
### What does this PR do?
Add approving or rejecting request feature
### Description of Task to be completed?

- Add route for approving or rejecting a request
- Add authorization middleware
- Add approving or rejecting request validations
- Add approving or rejecting request controller
- Add documentation for this route
- Add tests for this route
### How should this be manually tested?
```
$ git clone https://github.com/andela/the_spinners-backend.git
$ cd the_spinners-backend
$ git checkout  ft-reject-request-169817553
```
then 
`$ npm run test`

or you can use Postman to send patch request to /api/requests/:requestId with requestId as a paramenter and status which is 'approved' or 'rejected' as a request body.

### What are the relevant pivotal tracker stories?

#169817553
### Screenshots
Screen Shot 2020-01-10 at 15 30 05
Screen Shot 2020-01-10 at 15 32 28

Screen Shot 2020-01-11 at 01 20 15

### CODE REVIEW

File: src/controllers/request.controller.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class RequestController {
+  /**
+   *
+   *
+   * @static
+   * @param {req} req
+   * @param {res} res
+   * @returns {response} @memberof RequestController
+   */
+  static async rejectRequest(req, res) {
```
Comment (adesege) [inline src/controllers/request.controller.js]: You can make this method flexible by getting the request status from the request payload then set it here. This method can then be used to approve or reject a request.

File: src/controllers/request.controller.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import RequestService from '../services/request.service';
```
Comment (adesege) [inline src/controllers/request.controller.js]: I see there's a lot of similarities between this PR and #42. See the PR for the feedback I left there and implement them here.

Comment (gadishimwe) [reply src/controllers/request.controller.js]: Yeah, they are almost similar. ok let me do it.

**Review by adesege — APPROVED** [review]

File: src/validations/request.validation.js [hunk 3]
```diff
@@ -50,4 +51,37 @@ async function requestValidation(req, res, next) {
...
+      .messages({
+        'any.only': 'status" must be one of approved or rejected'
+      })
+      .required()
+  }).options({ abortEarly: false });
+
+  const results = schema.validate({ ...req.params, ...req.body });
+  if (results.error) {
+    const errorMessages = [];
+    results.error.details.forEach((error) => {
```
Comment (adesege) [inline src/validations/request.validation.js]: How many times more do I have to give feedback on the use of .map instead of .forEach?

Comment (gadishimwe) [reply src/validations/request.validation.js]: I'm sorry I missed it. it won't happen again. Thanks.

File: src/tests/fixtures/users.fixture.js [hunk 4]
```diff
@@ -86,6 +86,57 @@ export const loggedInToken = JwtService.generateToken({
   email: loggedInUser.email,
 });
 
+// create managers for management actions
+export const loggedInManager1 = {
```
Comment (adesege) [inline src/tests/fixtures/users.fixture.js]: Move this to a separate fixture file

File: src/tests/notification/notification.test.js [hunk 5]
```diff
@@ -93,4 +94,20 @@ describe('Tests notification preferences', () => {
         done(err);
       });
   });
+  it('Should return status code of 201 on successful trip creation', (done) => {
+    chai.request(app)
+      .post('/api/one-way-trips')
+      .set('Authorization', token)
+      .send({ ...newTrip, departureDate: '2020-05-23' })
```
Comment (adesege) [inline src/tests/notification/notification.test.js]: Use Fakerjs to autogenerate the date

File: src/tests/request/request.test.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import chai, { expect } from 'chai';
+import chaiHttp from 'chai-http';
+import app from '../../app';
+import createRequest from '../fixtures/request.fixture';
+import { managerToken1, managerToken2, NonManagerToken } from '../fixtures/users.fixture';
```
Comment (adesege) [inline src/tests/request/request.test.js]: Change the casing of `NonManagerToken` to camel case

File: src/tests/request/request.test.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+      .end((err, res) => {
+        expect(res).to.have.status(403);
+        expect(res.body).to.have.property('message');
+        expect(res.body.message).eqls('Forbidden. Only Managers can perform this action');
+        done();
+      });
+  });
+  it('Should return status code of 400 for invalid requestId parameter', (done) => {
+    chai.request(app)
+      .patch('/api/requests/gkkk')
```
Comment (adesege) [inline src/tests/request/request.test.js]: Use a random string to set the requestId

File: src/tests/request/request.test.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      .end((err, res) => {
+        expect(res).to.have.status(400);
+        expect(res.body).to.have.property('message');
+        expect(res.body.message[0]).eqls('requestId must be a number');
+        done();
+      });
+  });
+  it('Should return status code of 404 for requestId which does not exist', (done) => {
+    chai.request(app)
+      .patch('/api/requests/44')
```
Comment (adesege) [inline src/tests/request/request.test.js]: Use a random number to set the requestId.

File: src/tests/trip/comments.test.js [hunk 9]
```diff
@@ -9,18 +9,18 @@ chai.use(chaiHttp);
...
 describe('/POST create comment on trip request', () => {
   before(async () => {
-    await createTrip();
     await createUsers();
+    await createTrip();
   });
   it('Should allow user to create comments when provided successfully, userId, subjectId, subjectType and comment', (done) => {
     chai.request(app)
-      .post('/api/trip-requests/1/comments')
+      .post('/api/trip-requests/2/comments')
```
Comment (adesege) [inline src/tests/trip/comments.test.js]: Use a random number to set the requestId

File: src/services/request.service.js [hunk 10]
```diff
@@ -30,6 +30,33 @@ class RequestService {
...
     });
   }
+
+  /** find Request
+   * @static
+   * @param {object} property
+   * @memberof AccommodationService
+   * @returns {object} data
+   */
+  static findTypeByProperty(property) {
```
Comment (victor-abz) [inline src/services/request.service.js]: WHat if you rename it to findRequestByProperty as you are finding request

Comment (adesege) [reply src/services/request.service.js]: 👍

**Review by victor-abz — APPROVED** [review]
LGTM

**Pull request merged**